### PR TITLE
 *: importing data reduce memory allocations.

### PR DESCRIPTION
**Pull request by winkyao**
run  https://github.com/pingcap/tidb-lightning/tree/cenqichao/kv-encoder-bench with below command.

```
./bin/bench-encoder -r 200 -b 10240 -s 0
```
got below bench results:

```
master:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 93.126 sec

kv.ImportingTxnMembufCap = *batch * 32:

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 78.828 sec

kv.ImportingTxnMembufCap = *batch * 64:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 76.979 sec
```

about 17% improvement.

in flamegraph, we can see, after reuse the BufferStore in AddRecord, the memdb.New is optimized away. But memDB.Put --> runtime.growslice still use some times. If we wanna optimize it, we should reuse Bufferstore after we commit a transaction.

### CODE REVIEW

**Comment by winkyao** [comment]
[torch.svg.zip](https://github.com/pingcap/tidb/files/1600870/torch.svg.zip)

**Comment by shenli** [comment]
In [this PR](https://github.com/pingcap/tidb/pull/5546) I notice that there is a 20% improvement. So is it better that this PR?

**Comment by winkyao** [comment]
@shenli NO, this pr reuse the bs bufferstore, the bench results is mutable, so never mind.

**Comment by coocood** [comment]
How about that we add a `sync.Pool` for BufferStore instead of changing the argument of `AddRecord`?

**Comment by winkyao** [comment]
@shenli @coocood PTAL

**Comment by winkyao** [comment]
after reuse slice in tablecodec.EncodeRow, bench result:

```
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 73.352 sec
```
about 21% improvement.

File: table/tables/tables.go [hunk 1]
```diff
@@ -358,7 +358,7 @@ func (t *Table) AddRecord(ctx context.Context, r []types.Datum, skipHandleCheck
 	}
 
 	key := t.RecordKey(recordID)
-	value, err := tablecodec.EncodeRow(row, colIDs, ctx.GetSessionVars().GetTimeZone())
+	value, err := tablecodec.EncodeRow(row, colIDs, ctx.GetSessionVars().GetTimeZone(), ctx.GetSessionVars().RowValBuf[:0])
```
Comment (shenli) [inline table/tables/tables.go]: Maybe we should only reuse it when importing data. @coocood PTAL

**Comment by winkyao** [comment]
flamegraph: [torch.svg-kvencoder_optimize-5ad7bac16513.tar.gz](https://github.com/pingcap/tidb/files/1603949/torch.svg-kvencoder_optimize-5ad7bac16513.tar.gz)

**Comment by shenli** [comment]
@winkyao Please fix CI and resolve conflicts.

**Comment by winkyao** [comment]
@shenli @coocood PTAL

File: session.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ import (
 
 	"github.com/juju/errors"
 	"github.com/ngaut/pools"
-	"github.com/opentracing/opentracing-go"
+	opentracing "github.com/opentracing/opentracing-go"
```
Comment (coocood) [inline session.go]: Don't need to change.

File: session.go [hunk 3]
```diff
@@ -53,7 +53,7 @@ import (
 	"github.com/pingcap/tidb/util/auth"
 	"github.com/pingcap/tidb/util/charset"
 	"github.com/pingcap/tidb/util/kvcache"
-	"github.com/pingcap/tipb/go-binlog"
+	binlog "github.com/pingcap/tipb/go-binlog"
```
Comment (coocood) [inline session.go]: ditto

File: session.go [hunk 4]
```diff
@@ -878,10 +886,12 @@ func (s *session) NewTxn() error {
 			return errors.Trace(err)
 		}
 	}
+
```
Comment (coocood) [inline session.go]: remove the added empty lines in this PR.

File: executor/write.go [hunk 5]
```diff
@@ -806,6 +808,11 @@ func (e *LoadData) Open(goCtx goctx.Context) error {
 	return nil
 }
 
+type defaultVal struct {
+	val   types.Datum
+	valid bool
```
Comment (shenli) [inline executor/write.go]: Add comment for `valid`.

File: executor/write.go [hunk 6]
```diff
@@ -825,6 +832,9 @@ type InsertValues struct {
 
 	GenColumns []*ast.ColumnName
 	GenExprs   []expression.Expression
+
+	// colDefaultVals used to store casted default value.
```
Comment (shenli) [inline executor/write.go]: is used

File: executor/write.go [hunk 7]
```diff
@@ -839,13 +849,21 @@ type InsertExec struct {
...
 	// If tidb_batch_insert is ON and not in a transaction, we could use BatchInsert mode.
-	batchInsert := e.ctx.GetSessionVars().BatchInsert && !e.ctx.GetSessionVars().InTxn()
-	batchSize := e.ctx.GetSessionVars().DMLBatchSize
+	sessVars := e.ctx.GetSessionVars()
+	batchInsert := sessVars.BatchInsert && !sessVars.InTxn()
+	batchSize := sessVars.DMLBatchSize
 
 	txn := e.ctx.Txn()
 	rowCount := 0
+	tmpMemCap := 64
```
Comment (shenli) [inline executor/write.go]: Why not define it as a const?

File: sessionctx/variable/session.go [hunk 8]
```diff
@@ -248,6 +250,13 @@ type SessionVars struct {
 	// EnableChunk indicates whether the chunk execution model is enabled.
 	// TODO: remove this after tidb-server configuration "enable-chunk' removed.
 	EnableChunk bool
+
+	// RowValBuf used by tablecodec.EncodeRow, to reduce runtime.growslice.
```
Comment (shenli) [inline sessionctx/variable/session.go]: is used

File: sessionctx/variable/session.go [hunk 9]
```diff
@@ -248,6 +250,13 @@ type SessionVars struct {
...
+	RowValBuf []byte
+	// BufStore use to store temp kvs, to reduce memory allocations.
```
Comment (shenli) [inline sessionctx/variable/session.go]: use to store -> stores
I think the following comment is more clear:
```
BufStore stores temp KVs for a row when executing insert statement. We could reuse a BufStore for multiple rows of a session to reduce memory allocations.
```

File: table/tables/tables.go [hunk 10]
```diff
@@ -308,6 +310,14 @@ func (t *Table) rebuildIndices(ctx context.Context, rm kv.RetrieverMutator, h in
 	return nil
 }
 
+func adjustRowValuesBuf(sessVars *variable.SessionVars, rowLen int) {
```
Comment (shenli) [inline table/tables/tables.go]: Add comments for it.

File: table/tables/tables.go [hunk 11]
```diff
@@ -326,9 +336,13 @@ func (t *Table) AddRecord(ctx context.Context, r []types.Datum, skipHandleCheck
 	}
 
 	txn := ctx.Txn()
-	bs := kv.NewBufferStore(txn)
-
-	skipCheck := ctx.GetSessionVars().ImportingData
+	sessVars := ctx.GetSessionVars()
+	skipCheck := sessVars.ImportingData
```
Comment (shenli) [inline table/tables/tables.go]: Add comments for this logic. Shuaipeng Yu is confused with it.

File: tablecodec/tablecodec.go [hunk 12]
```diff
@@ -169,61 +170,64 @@ func EncodeValue(sc *stmtctx.StatementContext, raw types.Datum) ([]byte, error)
 
 // EncodeRow encode row data and column ids into a slice of byte.
 // Row layout: colID1, value1, colID2, value2, .....
-func EncodeRow(sc *stmtctx.StatementContext, row []types.Datum, colIDs []int64) ([]byte, error) {
+func EncodeRow(sc *stmtctx.StatementContext, row []types.Datum, colIDs []int64, valBuf []byte, values []types.Datum) ([]byte, error) {
```
Comment (shenli) [inline tablecodec/tablecodec.go]: Add comments for the two new arguments.

**Comment by shenli** [comment]
@coocood PTAL

**Comment by winkyao** [comment]
@shenli PTAL

File: executor/write.go [hunk 13]
```diff
@@ -806,6 +808,12 @@ func (e *LoadData) Open(goCtx goctx.Context) error {
 	return nil
 }
 
+type defaultVal struct {
+	val types.Datum
+	// valid indicates val is a casted default value.
```
Comment (shenli) [inline executor/write.go]: We evaluate the default value lazily. The `valid` indicates whether the `val` is evaluated.

File: util/kvencoder/kv_encoder.go [hunk 14]
```diff
@@ -226,6 +226,7 @@ func (e *kvEncoder) initial(dbName string, idAlloc autoid.Allocator) (err error)
 
 	se.GetSessionVars().IDAllocator = idAlloc
 	se.GetSessionVars().ImportingData = true
```
Comment (ngaut) [inline util/kvencoder/kv_encoder.go]: Why set the default ImportingData to true?

File: util/kvencoder/kv_encoder.go [hunk 15]
```diff
@@ -226,6 +226,7 @@ func (e *kvEncoder) initial(dbName string, idAlloc autoid.Allocator) (err error)
...
+	se.GetSessionVars().SkipUTF8Check = true
```
Comment (ngaut) [inline util/kvencoder/kv_encoder.go]: The same as ImportingData. Shouldn't it be set by the user session?

File: util/kvencoder/kv_encoder.go (hunk 14 above)
Comment (coocood) [reply util/kvencoder/kv_encoder.go]: The session is created and maintained by the kvEncoder.

File: util/kvencoder/kv_encoder.go (hunk 15 above)
Comment (coocood) [reply util/kvencoder/kv_encoder.go]: This session is only used by the kvEncoder.

**Pull request merged**